services/src/main/java/org/keycloak/keys/loader/PublicKeyStorageManager.java

```diff
@@ -46,6 +47,19 @@ public static PublicKey getClientPublicKey(KeycloakSession session, ClientModel
         return keyStorage.getPublicKey(modelKey, kid, loader);
     }
 
+    // KEYCLOAK-8460 client signed signature verification
+    // key is wrapped by KeyWrapper in order to use refactored token signature verification mechanism
+    public static KeyWrapper getClientPublicKeyWrapper(KeycloakSession session, ClientModel client, JWSInput input) {
+        String kid = input.getHeader().getKeyId();
+        PublicKey verifyKey = getClientPublicKey(session, client, input);
+        if (verifyKey == null) return null;
+
+        KeyWrapper keyWrapper = new KeyWrapper();
+        keyWrapper.setKid(kid);
+        keyWrapper.setAlgorithm(input.getHeader().getAlgorithm().name());
```

Comment:
I agree with. To do so with current client's public key management implementation (PublicKeyStorageManager, ClientPublicKeyLoader, InfinispanPublicKeyStorageProvider, PublicKeysEntry)
, I think that those might treat KeyWrapper (including key and its metadata) instead of PublicKey to retrieve the algorithm associated with the key. How do you think about it?

I've tried to retrieve and cache key information as KeyWrapper instead of PublicKey, but found that it needs to modify a lot of codes.

Reply:
I think that would be a great idea. Using KeyWrapper should be preferred over simply PublicKey as it allows setting the additional info that is needed such as kid and intended algorithm. It is after all quite important that the right keys are used for the right purpose.

Reply:
So, I will try to retrieve and cache the client public key information as KeyWrapper.
Current implementation gets the client public key from the following 3 sources.

* client's jwks_url location
* pre-registered certificate
* pre-registered public key

The first source might be tractable because I can map JWK formatted information onto KeyWrapper directly.
The second and third sources might not be tractable because no such metadata is registerd explicitly.
I've investigated them and found that keys from these two sources are for JWS Client Assertion for client authentication, and seemed to be that those key type is RSA, key use is for signature, algorithm is RS256 implicitly. Therefore, I will map keys' metadata which are from those two sources onto KeyWrapper as such. What do you think about it?

Reply:
With pre-registered certificate/public-key the ideal would be to add additional fields to set kid and algorithm. Not sure about migration of existing clients though, those could be set to RS256 and potentially kid could be generated as we do elsewhere when uploading a RSA public-key in PEM.

Reply:
To add, to fully support additional algorithms we should also add support to upload different keys as well.

Reply:
I agree with you. However, it might be difficult to incorporate this feature (register client public key/certificate with key type, key use, algorithm, etc explicitly) onto this PR.

In this time, in order to map pre-registered public key/certificate onto KeyWrapper, I would like to do the followings.

- key type is RSA
- key usage is signature
- algorithm is RS256